### PR TITLE
Add support types implementing IEquatable in Compare-Object cmdlet

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Address #12498

1. Currently Compare-Object is based on Compare() method pattern. The pattern is used for _ordering_  but the cmdlet must check _equality_. With the PR we implement Equal() methods and use them instead of Compare() methods.
2. This add support types implementing IEquatable interface.
3. As side effect performance of the cmdlet increases up to 35% (if compare Pester tests before and after the PR.)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: build.psm1 [hunk 1]
```diff
@@ -347,7 +347,9 @@ function Start-PSBuild {
         Write-Warning @"
```
Comment (PoshChan) [inline build.psm1:347]: @iSazonov, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Export-FormatData.Can export all types
```powershell
Expected $null, but got @(@{InputObject=Microsoft.PowerShell.Commands.X509StoreLocation; SideIndicator==>}, @{InputObject=System.Security.Cryptography.X509Certificates.X509Certificate2; SideIndicator==>}, @{InputObject=System.Security.Cryptography.X509Certificates.X509Store; SideIndicator==>}, @{InputObject=System.CodeDom.Compiler.CompilerError; SideIndicator==>}, @{InputObject=System.Reflection.Assembly; SideIndicator==>}, @{InputObject=System.Reflection.AssemblyName; SideIndicator==>}, @{InputObject=System.Globalization.CultureInfo; SideIndicator==>}, @{InputObject=System.Diagnostics.FileVersionInfo; SideIndicator==>}, @{InputObject=System.Diagnostics.EventLogEntry; SideIndicator==>}, @{InputObject=System.Diagnostics.EventLog; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-FormatData.Tests.ps1: line 30
30:             Compare-Object $fd $actualAllFormat | Should -Be $null
```
Select-String.String actions.Should use the ca alias for casesensitive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1: line 48
48:             $equal | Should -BeTrue
```
Stream writer tests.Write-Information cmdlet.Write-Information works with defaults
```powershell
Expected 0, but got 2.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Stream.Tests.ps1: line 115
115:             (Compare-Object $result $returnValue -SyncWindow 0).length | Should -Be 0
```
Stream writer tests.Write-Information cmdlet.Write-Information works with -Object
```powershell
Expected 0, but got 2.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Stream.Tests.ps1: line 115
115:             (Compare-Object $result $returnValue -SyncWindow 0).length | Should -Be 0
```
Stream writer tests.Write-Information cmdlet.Write-Information works with -Message
```powershell
Expected 0, but got 2.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Stream.Tests.ps1: line 115
115:             (Compare-Object $result $returnValue -SyncWindow 0).length | Should -Be 0
```

**Comment by iSazonov** [comment]
@mklement0 Could you please review and look CI failures? Is there something we need to worry about?

File: build.psm1 (hunk 1 above)
Comment (PoshChan) [inline build.psm1:347]: @iSazonov, your last commit had 9 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Get-Content.Check Get-Content containing multi-byte chars.Should return last three lines two lines at a time for -ReadCount 2 and -Tail 3
```powershell
Expected $null or empty, but got @(@{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator=<=}, @{InputObject=System.Object[]; SideIndicator=<=}).
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1: line 341
341:             Compare-Object -ReferenceObject $expected -DifferenceObject $result | Should -BeNullOrEmpty
```
Get-Content.Check Get-Content containing multi-byte chars.Should return first three lines two lines at a time for -TotalCount 3 and -ReadCount 2
```powershell
Expected $null or empty, but got @(@{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator=<=}, @{InputObject=System.Object[]; SideIndicator=<=}).
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1: line 353
353:             Compare-Object -ReferenceObject $expected -DifferenceObject $result | Should -BeNullOrEmpty
```
Export-FormatData.Can export all types
```powershell
Expected $null, but got @(@{InputObject=Microsoft.PowerShell.Commands.X509StoreLocation; SideIndicator==>}, @{InputObject=System.Security.Cryptography.X509Certificates.X509Certificate2; SideIndicator==>}, @{InputObject=System.Security.Cryptography.X509Certificates.X509Store; SideIndicator==>}, @{InputObject=System.CodeDom.Compiler.CompilerError; SideIndicator==>}, @{InputObject=System.Reflection.Assembly; SideIndicator==>}, @{InputObject=System.Reflection.AssemblyName; SideIndicator==>}, @{InputObject=System.Globalization.CultureInfo; SideIndicator==>}, @{InputObject=System.Diagnostics.FileVersionInfo; SideIndicator==>}, @{InputObject=System.Diagnostics.EventLogEntry; SideIndicator==>}, @{InputObject=System.Diagnostics.EventLog; SideIndicator==>}).
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-FormatData.Tests.ps1: line 30
30:             Compare-Object $fd $actualAllFormat | Should -Be $null
```
Select-String.String actions.Should use the ca alias for casesensitive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1: line 48
48:             $equal | Should -BeTrue
```
Stream writer tests.Write-Information cmdlet.Write-Information works with defaults
```powershell
Expected 0, but got 2.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Stream.Tests.ps1: line 115
115:             (Compare-Object $result $returnValue -SyncWindow 0).length | Should -Be 0
```

Comment (PoshChan) [inline build.psm1:347]: @iSazonov, your last commit had 14 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Get-Content.Check Get-Content containing multi-byte chars.Should return last three lines two lines at a time for -ReadCount 2 and -Tail 3
```powershell
Expected $null or empty, but got @(@{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator=<=}, @{InputObject=System.Object[]; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Get-Content.Tests.ps1: line 341
341:             Compare-Object -ReferenceObject $expected -DifferenceObject $result | Should -BeNullOrEmpty
```
Get-Content.Check Get-Content containing multi-byte chars.Should return first three lines two lines at a time for -TotalCount 3 and -ReadCount 2
```powershell
Expected $null or empty, but got @(@{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator==>}, @{InputObject=System.Object[]; SideIndicator=<=}, @{InputObject=System.Object[]; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Get-Content.Tests.ps1: line 353
353:             Compare-Object -ReferenceObject $expected -DifferenceObject $result | Should -BeNullOrEmpty
```
Get-Service cmdlet tests.Get-Service filtering works for '{ Get-Service -DisplayName Net* }'
```powershell
Expected $null or empty, but got 
InputObject       SideIndicator
-----------       -------------
NcaSvc            =>
NcbService        =>
Netlogon          =>
Netman            =>
NetMsmqActivator  =>
NetPipeActivator  =>
netprofm          =>
NetSetupSvc       =>
NetTcpActivator   =>
NetTcpPortSharing =>
NlaSvc            =>
nsi               =>
NcaSvc            <=
NcbService        <=
Netlogon          <=
Netman            <=
NetMsmqActivator  <=
NetPipeActivator  <=
netprofm          <=
NetSetupSvc       <=
NetTcpActivator   <=
NetTcpPortSharing <=
NlaSvc            <=
nsi               <=

.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Get-Service.Tests.ps1: line 69
69:       Compare-object $services $servicesCheck | Out-String | Should -BeNullOrEmpty
```
Get-Service cmdlet tests.Get-Service filtering works for '{ Get-Service -Include Net* -Exclude *logon }'
```powershell
Expected $null or empty, but got 
InputObject       SideIndicator
-----------       -------------
Netman            =>
NetMsmqActivator  =>
NetPipeActivator  =>
netprofm          =>
NetSetupSvc       =>
NetTcpActivator   =>
NetTcpPortSharing =>
Netman            <=
NetMsmqActivator  <=
NetPipeActivator  <=
netprofm          <=
NetSetupSvc       <=
NetTcpActivator   <=
NetTcpPortSharing <=

.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Get-Service.Tests.ps1: line 69
69:       Compare-object $services $servicesCheck | Out-String | Should -BeNullOrEmpty
```
Get-Service cmdlet tests.Get-Service filtering works for '{ Get-Service -Name Net* | Get-Service }'
```powershell
Expected $null or empty, but got 
InputObject       SideIndicator
-----------       -------------
Netlogon          =>
Netman            =>
NetMsmqActivator  =>
NetPipeActivator  =>
netprofm          =>
NetSetupSvc       =>
NetTcpActivator   =>
NetTcpPortSharing =>
Netlogon          <=
Netman            <=
NetMsmqActivator  <=
NetPipeActivator  <=
netprofm          <=
NetSetupSvc       <=
NetTcpActivator   <=
NetTcpPortSharing <=

.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Get-Service.Tests.ps1: line 69
69:       Compare-object $services $servicesCheck | Out-String | Should -BeNullOrEmpty
```

**Comment by iSazonov** [comment]
@mklement0 Could you please review and look CI failures?

Original behavior is that if we have two not comparable objects as last resort we convert to string and compare the strings.
Now we call Equal() and return result right away without the last resort conversion to strings. And it is a breaking change.

**Comment by vexx32** [comment]
@iSazonov is doing a string conversion that expensive? Perhaps it would be better to leave that in as a fallback check? Or provide a switch parameter to use that mode?

**Comment by iSazonov** [comment]
> Perhaps it would be better to leave that in as a fallback check?

I do not know if this is possible for Equals(). Need to analyze more.
In "compare" pattern we use
`        public static bool TryCompare(object first, object second, bool ignoreCase, IFormatProvider formatProvider, out int result)`
I am not sure that we can use the logic for "equals".

File: build.psm1 (hunk 1 above)
Comment (PoshChan) [inline build.psm1:347]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-linux`
Language Primitive Tests.Equality comparison with string and non-numeric type should not be culture sensitive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Api/LanguagePrimitive.Tests.ps1: line 7
7:         $val | Should -BeTrue
```

Comment (PoshChan) [inline build.psm1:347]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
Language Primitive Tests.Equality comparison with string and non-numeric type should not be culture sensitive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/engine/Api/LanguagePrimitive.Tests.ps1: line 7
7:         $val | Should -BeTrue
```

Comment (PoshChan) [inline build.psm1:347]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
Language Primitive Tests.Equality comparison with string and non-numeric type should not be culture sensitive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Api\LanguagePrimitive.Tests.ps1: line 7
7:         $val | Should -BeTrue
```

**Pull request closed without merging**